src/runtime/stm32/graph_runtime.cc

```diff
@@ -0,0 +1,529 @@
+/*
```

Comment:
Thanks Andrew, @areusch 
I am aligning on your comments.

Reply:
How are you, Andrew @areusch

I have reorganized the code like so:

1. Removed the `apps/stm32`
2. Removed linker script code from the `emitter.py`
3. Removed the copy of `graph_runtime` from `src/runtime/stm32`
3. Moved the `src/runtime/stm32` to `src/runtime/crt/contrib/stm32`. I have chosen the `contrib' because it looks like other files in the directory are sort of common implementation, while the `stm32` is code-generator specific. If you prefer, not to use `contrib` for this - I will align the runtime and tests.
4. Moved the C part of our CI test to `tests/crt/contrib/stm32`

Now, I am looking at this comment:

> it seems like you could use the TVM C runtime's crt_backend_api.c and implement TVMPlatformMemoryAllocate(). Curious if that would be an option for you? btw, it would be totally fine with me if you wanted to split this file into a separate CRT library e.g. src/runtime/crt/backend to avoid including the rest of src/runtime/crt/common.

I have two observations:

1. Current TVM generates C code with these calls:
```
TVMAPISetLastError()
TVMGetLastError()
TVMBackendAllocWorkspace()
TVMBackendFreeWorkspace()
```

These would be good candidates for including with the `crt_backend_api.c`. Therefore, moving the `TVMAPISetLastError()` and the `TVMGetLastError()` there from the `crt_runtime_api.c`, and moving the `TVMBackendParallelLaunch()`, `TVMBackendRegisterSystemLibSymbol()` out to `crt_runtime_api.c`.

This way, the `crt_backend_api.c` would carry a minimal support required by the C target code generator.

2. Currently, the `TVMBackendAllocWorkspace` does some error handling. I would propose to leave the error handling to the caller, while the `TVMBackendAllocWorkspace` would just return `NULL` in case of error.
The reason is that the error handling and logging are platform-specific and even application-specific - eg. `malloc` does not do error handling, for example.
Handling this in the `TVMBackendAllocWorkspace` requires implementing the `TVMLogf` and the `TVMPlatformAbort`, which we will not be able to test. For example, in our `stm32` TVM test, I would have to implement the `x86` specific error handling, while for the Disco board I would set it to NOOP since the applications handles errors in its own way.

3. If we were to reduce the `crt_backend_api.c` to its minimal form as I have described above, perhaps it makes sense to have the application implement these few methods directly, i.e. no need to include the `tvm/crt/crt_backend_api.c` implementation, only the `crt_backend_api.h` header as interface ?

What do you think ?

Do not review the `tests/crt/contrib/stm32` yet - I will align it and clean after I have your opinion on the above.

cheers
Arthur

Reply:
hi @stoa, sorry for the delay again. i'll take a look at the parts of this PR outside of `tests/crt/contrib/stm32`.

> These would be good candidates for including with the crt_backend_api.c. Therefore, moving the TVMAPISetLastError() and the TVMGetLastError() there from the crt_runtime_api.c, and moving the TVMBackendParallelLaunch(), TVMBackendRegisterSystemLibSymbol() out to crt_runtime_api.c.

I think that makes sense to me. `crt_backend_api.c` should be able to be largely reused across platforms.

> Currently, the TVMBackendAllocWorkspace does some error handling. I would propose to leave the error handling to the caller, while the TVMBackendAllocWorkspace would just return NULL in case of error.
The reason is that the error handling and logging are platform-specific and even application-specific - eg. malloc does not do error handling, for example.

There's a couple concerns I have here:
1. We'd have to generate error-checking code in operator implementations, since that is where `TVMBackendAllocWorkspace` is called from. It's hard to generate macros such as `CHECK_EQ` there.
2. Returning NULL would obscure the error code. I added the error code for traceability--in my experience, it's often valuable to have a trace of the execution path that trips an error code. On embedded, it can be quite hard to get these traces particularly in production. Therefore, having the error code helps to debug without necessarily needing to have a debugger attached. We could explore changing the signature of TVMBackendAllocWorkspace, but it could break people

> If we were to reduce the crt_backend_api.c to its minimal form as I have described above, perhaps it makes sense to have the application implement these few methods directly, i.e. no need to include the tvm/crt/crt_backend_api.c implementation, only the crt_backend_api.h header as interface ?

Potentially for the memory allocation function, but I'm not so sure about the rest. The rest are for interop with the runtime. As we add support for e.g. accelerators or better memory planning here, I could see these implementations becoming a bit more complex. I'd prefer to stick with the application hook `TVMPlatformMemoryAllocate` for now, if that's the only divergence you can see per-application.

Reply:
Hello, Andrew @areusch 

My concern is imposing to implement the TVMLogf and the TVMPlatformAbort. While we implement the TVMPlatformAlloc as part of the stm32 runtime (this is a fairly common functionality across various boards), the TVMLogf and the TVMPlatformAbort are different. They are really board and even application-specific. We would like to avoid having hooks from the runtime back to the application. Returning the error codes from the crt_backend_api seems to be our commonly preferred way of handling errors here - I am very interested in this being proposed for the mTVM.

> We'd have to generate error-checking code in operator implementations, since that is where TVMBackendAllocWorkspace is called from. It's hard to generate macros such as CHECK_EQ there.

I am not sure I understand this point. If the memory allocation fails, we have 2 choices:
 - The operator code catches it and SetLastError(), eventually with a detailed error code. I do not see why this is difficult or not clean - but perhaps I do not have enough vision here. In any case, this seems to have already been implemented with the TVM operators, Is this going to change ? Which RFC ?
 - The crt layer catches the error and handles via further hooks into the application code. We prefer to minimize such hooks.
I am not sure why you have mentioned the CHECK_EQ macro ?

So, I am not sure how I should proceed with the test. I could do this:

1. I move the Set/GetLatError from the crt_runtime_api.c to the crt_backend_api.c.
2. We include the crt_backend_api.c with our test. The TVMLogf and the TVMPlatformAbort will be implemented 'dummy' with the stm32 runtime.

Is this gonna work for you ?

Reply:
hi Arthur @stoa,

Ah I think I understand what's going on here...sorry for all of this confusion. You are stumbling into some problems we have not yet addressed for production inference use cases. There are a couple of things at play here:

### Error-handling paradigm and location of `TVMAPISetLastError()`.

`TVMAPISetLastError` and `TVMGetLastError` should be in `crt_backend_api.h` as you pointed out. However, microTVM has meanwhile adopted `tvm_crt_error_t` for many of its non-c_backend_api (and indeed, also in c_backend_api via calls to TVMPlatform*) functions. 

> Returning the error codes from the crt_backend_api seems to be our commonly preferred way of handling errors here - I am very interested in this being proposed for the mTVM.

I concur with this opinion and this is what I'm going for with `tvm_crt_error_t`. I hadn't realized the misplacement of `TVMAPISetLastError`, since (from my reading) it's actually only used in generated code to handle assert fails, and I don't think we often use those in microTVM. Given this, I think it would be worth clarifying error handling in microTVM with a separate RFC. I don't think we should tackle that in this PR--perhaps a separate PR could:
1. Move `TVMAPISetLastError` and `TVMGetLastError` to `c_backend_api.h`.
2. Propose use of `tvm_crt_error_t` as the return value from TVM API functions, at least in the C runtime API. We could take this a step further and replace the `int` return value in the C++ runtime, but it may get complicated and `tvm_crt_error_t` is indeed compatible with the zero-indicates-success paradigm currently mandated by `c_{backend,runtime}_api.h`
3. Allocate a `tvm_crt_error_t` code to indicate assert failure, and direct users to `TVMGetLastError` to retrieve the error text.

I am happy to take on this RFC/PR if you would be okay waiting on it; or if you want to do it, I'm completely fine with that. Let me know if this direction makes sense to you.

###  Handling `TVMBackendAllocWorkspace` errors

> I am not sure I understand this point. If the memory allocation fails, we have 2 choices:
> 
> - The operator code catches it and SetLastError(), eventually with a detailed error code. I do not see why this is difficult or not clean - but perhaps I do not have enough vision here. In any case, this seems to have already been implemented with the TVM operators, Is this going to change ? Which RFC ?
> - The crt layer catches the error and handles via further hooks into the application code. We prefer to minimize such hooks.
> I am not sure why you have mentioned the CHECK_EQ macro ?

Here I was examining the existing implementation of `TVMBackendAllocWorkspace` in `src/runtime/crt/common/crt_backend_api.c`. That function is currently required to return NULL on error, so we would need to either:
1. change it to return `int` or `tvm_crt_error_t` to indicate errors
2. Make it follow the same paradigm as assert fails above

I prefer 1 here, since an error code can programmatically indicate what's wrong and is easier to log. Perhaps it's best to propose this as part of the RFC mentioned above.

I mentioned `CHECK_EQ` because that is the existing error-handling strategy. I think this calls `TVMPlatformAbort`, so I'm not sure it's the right one for inference deployment.

### Implementation of `TVMLogf` and `TVMPlatformAbort`

I agree at least `TVMLogf` should be optional in deployment. Likely, so should `TVMPlatformAbort`, though I think it remains to be seen how we can handle errors in higher-level libraries. I don't see operator code as needing to call `TVMPlatformAbort`, unless accelerator or other libraries need such a facility.

In any case, I agree it's fine to implement these as dummy for your production inference STM32 applications--once we resolve the `TVMBackendAllocWorkspace` error-handling strategy question above.

cc @tqchen regarding the API problems.